Another basis set for multi electron on one site

liwt31 opened the pull request:
Solves #89 .
Some points worth noting:
-  `Holstein_model_scheme123` and `Holstein_model_scheme4` in `ModelTranslator` are removed because in `MolList2` these schemes are firstly transformed into general model (or vibronic model) for future MPO construction. 
- The new basis is designed to work with multiple multi-e bases in the same model. The feature has already been put into use for a ladder-shaped Peierls Hamiltonian. This is why `dof_idx_map` is required for the basis.
- The special logic for reduced density matrix calculation when multiple e dofs are on the same site is simplified owing to the new expectation calculation routine #84 .

jjren commented on renormalizer/model/mlist.py:
```diff
@@ -351,10 +337,10 @@ def e_idx(idx):
             elif mol_list.scheme == 4:
-                dipole[(f"e_{e_idx(imol)}", "e_0")] = mol.dipole
-                dipole[("e_0", f"e_{e_idx(imol)}")] = mol.dipole
+                dipole[(f"e_{imol}", "e_0")] = mol.dipole
```
It is not correct here. The dipole matrix is between vacuum and e_{imol}. 
I think we can only keep `dipole[(f"e_{imol}", )] = mol.dipole` now and come back again when we deal with this kind of "transition" problem someday.

jjren commented on renormalizer/mps/mps.py:
```diff
@@ -1495,50 +1495,26 @@ def full_wfn(self) -> np.array:
+                                              model_translator=ModelTranslator.general_model)
+                        mpos.append(mpo)
+                self.mol_list.mpos[key] = mpos
             else:
-                e_idx = self.mol_list.multi_e_idx
-                l = ones((1, 1))
-                for i in range(e_idx):
-                    mat = self[i].reshape(self[i].shape[0], -1, self[i].shape[-1])
-                    l = tensordot(l, mat, axes=(0, 0))
-                    l = tensordot(l, mat.conj(), axes=([0, 1], [0, 1]))
-                r = ones((1, 1))
-                for i in range(len(self)-1, e_idx, -1):
-                    mat = self[i].reshape(self[i].shape[0], -1, self[i].shape[-1])
-                    r = tensordot(mat, r, axes=(-1, 0))
-                    r = tensordot(r, mat.conj(), axes=([1, 2], [1, 2]))
```
If this part is deleted,  the `BasisMultiElectron` class can not calculate the reduced density matrix anymore.

jjren commented on renormalizer/utils/basis.py:
```diff
@@ -150,8 +155,10 @@ def op_mat(self, op):
+            if op_symbol[0] == "I":
+                mat = np.eye(self.nbas)
+            elif op_symbol[0] == "a" or op_symbol[0] == r"a^\dagger":
+                raise ValueError(f"op_symbol:{op_symbol} is not supported. Try use BasisMultiElectronVac.")
```
add `else` is better.

liwt31 commented on renormalizer/utils/basis.py:
```diff
@@ -176,6 +179,75 @@ def op_mat(self, op):
+                if op_symbol_term == r"a^\dagger":
+                    mat[op_symbol_idx, 0] = 1.
+                elif op_symbol_term == r"a":
+                    mat[0, op_symbol_idx] = 1.
```
This is done in line 246. It is ensured a valid operator is returned.

jjren replied:
I do not think so. If an unsupported operator is met, `mat` will be `np.zeros((self.nbas, self.nbas))`. Right?

jjren commented:
A good idea to add `dof_idx_map` to allow more than one multi-e sites in a DMRG chain.
I believe it would be helpful to reduce the MPO bond dimension of a ladder Hamiltonian.
It is better to add a test for this feature.

liwt31 commented:
I've fixed the bug and added a test as required. 
> If this part is deleted, the BasisMultiElectron class can not calculate the reduced density matrix anymore.

I'm not sure I understand what this means exactly, but according to the newly added test it's not a problem to calculate reduced density matrix for BasisMultiElectron.

jjren commented:
> I've fixed the bug and added a test as required.
> 
> > If this part is deleted, the BasisMultiElectron class can not calculate the reduced density matrix anymore.
> 
> I'm not sure I understand what this means exactly, but according to the newly added test it's not a problem to calculate reduced density matrix for BasisMultiElectron.

I'm wrong. Forget about it.
